src/server/dragonfly_test.cc

```diff
@@ -758,21 +758,72 @@ TEST_F(DflyEngineTest, Bug496) {
+  const uint64_t kMinExpectedDefragCount = 80;
+
+  std::vector<std::string> keys2delete;
+  keys2delete.push_back("del");
+
+  // Generate a list of keys that would be deleted
+  // The keys that we will delete are all in the form of "key-name:1<other digits>"
+  // This is because we are populating keys that has this format, but we don't want
+  // to delete all keys, only some random keys so we deleting those that start with 1
+  constexpr int kFactor = 10;
+  int kMaxNumKeysToDelete = 10'000;
+  int current_step = kFactor;
+  for (int i = 1; i < kMaxNumKeysToDelete; current_step *= kFactor) {
+    for (int j = i - 1 + current_step; i < current_step; i++, j++) {
+      keys2delete.push_back("key-name:" + std::to_string(j));
+    }
+  }
+
+  std::vector<std::string_view> keys(keys2delete.begin(), keys2delete.end());
 
-  RespExpr resp = Run({"DEBUG", "POPULATE", std::to_string(NUMBER_OF_KEYS), "key-name", "130"});
+  RespExpr resp = Run(
+      {"DEBUG", "POPULATE", std::to_string(kNumberOfKeys), "key-name", std::to_string(kKeySize)});
   ASSERT_EQ(resp, "OK");
   resp = Run({"DBSIZE"});
-  EXPECT_THAT(resp, IntArg(NUMBER_OF_KEYS));
+  EXPECT_THAT(resp, IntArg(kNumberOfKeys));
 
   shard_set->pool()->AwaitFiberOnAll([&](unsigned index, ProactorBase* base) {
     EngineShard* shard = EngineShard::tlocal();
     ASSERT_FALSE(shard == nullptr);  // we only have one and its should not be empty!
     EXPECT_EQ(shard->GetDefragStats().success_count, 0);
-    // we are not running stats yet
-    EXPECT_EQ(shard->GetDefragStats().tries, 0);
-    EXPECT_GT(GetMallocCurrentCommitted(), NUMBER_OF_KEYS);
+    // we are expecting to have at least one try by now
+    EXPECT_GT(shard->GetDefragStats().tries, 0);
+    EXPECT_GT(GetMallocCurrentCommitted(), kMaxNumKeysToDelete);
+  });
+
+  ArgSlice delete_cmd(keys);
+  auto r = CheckedInt(delete_cmd);
+  // the first element in this is the command del so size is one less
+  ASSERT_EQ(r, keys2delete.size() - 1);
+
+  // At this point we need to see whether we did running the task and whether the task did something
+  shard_set->pool()->AwaitFiberOnAll([&](unsigned index, ProactorBase* base) {
+    EngineShard* shard = EngineShard::tlocal();
+    ASSERT_FALSE(shard == nullptr);  // we only have one and its should not be empty!
+    int tries = 10;
+    uint64_t defrag_count = base->AwaitBrief([&]() {
+      // This task is kind of condition to await on for the scan to run and complete for at least
+      // one time. Since there is no option to sleep (it would just block all fibers), we are
+      // creating a busy wait fiber here that checks the stats whether the defrag task run and
+      // whether it did something.
+      auto stats = shard->GetDefragStats();
+      while (tries > 0 && stats.tries < 3) {
+        stats = shard->GetDefragStats();
+        if (stats.success_count > 10) {
+          return stats.success_count;
+        }
+        --tries;
+        this_fiber::sleep_for(220ms);
```

Comment:
what do you want to test in words?

Reply:
We need to somehow make sure that the idle task was executed and that it found a memory to defrag (it should once it was running since we deleted keys at random locations). I have to say that this didn't crashed, but it makes the test a little bit unpredictable in the sense that you cannot be sure whether the task ran or not. I though of adding a "TEST" functions that would be called explicitly to execute the DoDefarg, but didn't like the idea of adding functions that are should not be part of the production code.